### PR TITLE
Port backtrack tactic from Mathlib

### PR DESCRIPTION
**Pull request by joehendrix**
This is to support library_search port.

### CODE REVIEW

File: Std/Control/Nondet/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+/--
+Lift a list of values to a nondeterministic value.
+(The backtrackable state in each will be identical:
+whatever the state was when we first read from the result.)
+-/
+def ofList (L : List α) : Nondet m α := ofListM (L.map pure)
+
+/-- Apply a function which returns values in the monad to every alternative of a `Nondet m α`. -/
+partial def mapM (f : α → m β) (L : Nondet m α) : Nondet m β :=
```
Comment (semorrison) [inline Std/Control/Nondet/Basic.lean:123]:
```suggestion
def mapM (f : α → m β) (L : Nondet m α) : Nondet m β :=
```

File: Std/Control/Nondet/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+def ofOptionM (x : m (Option α)) : Nondet m α := .squash fun _ => do
+  match ← x with
+  | none => return .nil
+  | some a => return singleton a
+
+/-- Convert an optional value to a nondeterministic value. -/
+def ofOption (x : Option α) : Nondet m α := ofOptionM (pure x)
+
+/-- Filter and map a nondeterministic value using a monadic function which may return `none`. -/
+partial def filterMapM (f : α → m (Option β)) (L : Nondet m α) : Nondet m β :=
```
Comment (semorrison) [inline Std/Control/Nondet/Basic.lean:140]:
```suggestion
def filterMapM (f : α → m (Option β)) (L : Nondet m α) : Nondet m β :=
```

File: Std/Control/Nondet/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  L.bind fun a => ofOptionM (f a)
+
+/-- Filter and map a nondeterministic value. -/
+def filterMap (f : α → Option β) (L : Nondet m α) : Nondet m β :=
+  L.filterMapM fun a => pure (f a)
+
+/-- Filter a nondeterministic value using a monadic predicate. -/
+partial def filterM (p : α → m (ULift Bool)) (L : Nondet m α) : Nondet m α :=
```
Comment (semorrison) [inline Std/Control/Nondet/Basic.lean:148]:
```suggestion
def filterM (p : α → m (ULift Bool)) (L : Nondet m α) : Nondet m α :=
```

File: Std/Control/Nondet/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  toMLList : MLList m (α × σ)
+
+namespace Nondet
+
+variable {m : Type → Type} [Monad m] [MonadBacktrack σ m]
+
+/-- The empty nondeterministic value. -/
+def nil : Nondet m α := .mk .nil
+
+instance : Inhabited (Nondet m α) := ⟨.nil⟩
```
Comment (semorrison) [inline Std/Control/Nondet/Basic.lean:59]:
```suggestion
instance : EmptyCollection (Nondet m α) := ⟨.nil⟩

instance : Inhabited (Nondet m α) := ⟨.nil⟩
```

**Review by semorrison — APPROVED** [review]

**Comment by joehendrix** [comment]
Removing in favor of just adding this as part of solve_by_elim.

**Pull request closed without merging**